### ISSUE TITLE
vmalert memory usage increases since v1.104.0

### ISSUE THREAD

**Issue by andersbulderbank** [VictoriaMetrics/VictoriaMetrics#7471]
### Describe the bug

We are seeing memory usage for vmalert increasing over time in our setup, until Kubernetes OOM kills it because of exeeded memory limits. Looking at our metrics, this seems to have started in the beginning of October, perhaps coinciding with upgrading vmalert helm chart from v0.11.1 to v0.12.0 (October 3rd), and then worsening around the same time that we upgraded from v0.12.0 to v0.12.1 (October 7th):

![image](https://github.com/user-attachments/assets/79355cc7-0a7b-4fad-a2b3-425b3c1eb866)

### To Reproduce

Our configuration has 47 rules, most are pretty standard alerts for Kubernetes and VictoriaMetrics, and a few are domain specific, a few examples here, I can post the complete alert-config if you wish:
```
groups:
  - name: backend-events
    interval: 1m
    labels:
      team: backend
      severity: warning
    rules:
      # Record query results for efficiency / making alerts more DRY
      - record: event:start:count:30m
        expr: sum by (app, event) (increase(backend_event_total{status = "start"}[30m]))
      - record: event:fail:count:30m
        expr: sum by (app, event) (increase(backend_event_total{status = "fail"}[30m]))
      - record: event:fail:count:10m
        expr: sum by (app, event) (increase(backend_event_total{status = "fail"}[10m]))

      # Alerts for all events
      - alert: HighEventFailureRate
        for: 1s
        expr: |
          WITH (
            startCount = event:start:count:30m,
            failCount = event:fail:count:30m,
            minFailRate = 50,
            minEvents = 5
          )
          ( failCount / startCount ) * 100 >= minFailRate and startCount >= minEvents
        annotations:
          summary: 'High event failure rate(s):'
          description: '{{ $labels.event }} ({{ $labels.app }}): {{ $value | printf "%.2f" }}% failure rate over the last 30m.'
```
```
      - name: containers
        interval: 1m
        labels:
          team: cloud
          severity: warning
        rules:
          - alert: AppHighMemUtilization
            for: 10m
            expr: |
              WITH(
                filters = {namespace=~"backend-apps|web-apps"},
                requested = sum(eagle_pod_container_resource_requests_memory_bytes{filters}) by (container),
                consumed = sum(eagle_pod_container_resource_usage_memory_bytes{filters}) by (container),
                utilization = (consumed / requested) and requested != 0
              )
              avg(utilization) by (container) * 100 > 85
            annotations:
              summary: 'High container memory utilization for Bulder app, check if autoscaling is working:'
              description: '{{ $labels.container }}/{{ $labels.container }}:  {{ $value | printf "%.0f" }}% of requested for over 10m.'
```
Our helm values: 
```
    releaseName: victoria-metrics-alert
    version: 0.12.3
    repo: https://victoriametrics.github.io/helm-charts/
    valuesInline:
      fullnameOverride: vmalert
      server:
        fullnameOverride: vmalert
        configMap: vmalert-alerts
        extraArgs:
          rule: "/config/alerts-*.yaml"
        remote:
          read:
            url: http://vmsingle.monitoring.svc.cluster.local:8428
          write:
            url: http://vmsingle.monitoring.svc.cluster.local:8428
        datasource:
          url: http://vmsingle.monitoring.svc.cluster.local:8428
        notifier:
          alertmanager:
            url: http://alertmanager.monitoring.svc.cluster.local:9093
        resources:
          requests:
            cpu: 3m
            memory: 128Mi
          limits:
            memory: 128Mi
```

### Version

We are running victoriametrics/vmalert:v1.106.0 docker image at this time, but as mentioned earlier, we have upgraded a few times since this memory behaviour started appearing.

```
~ $ /vmalert-prod --version
vmalert-20241104-144158-tags-v1.106.0-0-g371e193279
```

### Logs

We have not seen any signs of errors or warnings that seems applicable to this issue.

### Screenshots

![31a5a54d01b946dcb4748019b5a18aca](https://github.com/user-attachments/assets/485360a0-5557-444e-af70-cbae01954d77)

### Used command-line flags

```
--datasource.url=http://vmsingle.monitoring.svc.cluster.local:8428
--envflag.enable=true
--envflag.prefix=VM_
--loggerFormat=json
--notifier.url=http://alertmanager.monitoring.svc.cluster.local:9093
--remoteRead.url=http://vmsingle.monitoring.svc.cluster.local:8428
--remoteWrite.url=http://vmsingle.monitoring.svc.cluster.local:8428
--rule=/config/alerts-*.yaml
```

### Additional information

Memory profile of vmalert 30 hours apart
[2024-11-04-1210-mem.zip](https://github.com/user-attachments/files/17658977/2024-11-04-1210-mem.zip)
[2024-11-05-1810-mem.zip](https://github.com/user-attachments/files/17658979/2024-11-05-1810-mem.zip)

**Comment by andersbulderbank** [comment]
Hi @Haleygo, see attached screenshot
![image](https://github.com/user-attachments/assets/009e1db3-6bb9-4237-adfe-5692abade6c6)

**Comment by andersbulderbank** [comment]
@Haleygo BTW, here is a screenshot of a longer period, regarding the dates we upgraded vmwagent to a new version.
![image](https://github.com/user-attachments/assets/e2458a63-6857-4243-905f-fa2de7834b01)

**Comment by Haleygo** [comment]
Hi @andersbulderbank ,
As I can see, the memory limit&request for vmalert are both 128Mi, but in the graph you posted in description, vmalert restarted at around 80%, is that caused by other external tool?
Could you try increasing the memory limit a bit more(like 200Mi) to prevent vmalert from crashing until the memory stabilizes? And try increasing `gogc` to 100 to see if that helps stabilize memory more quickly? The value for helm would be:
```
    releaseName: victoria-metrics-alert
    version: 0.12.3
    repo: https://victoriametrics.github.io/helm-charts/
    valuesInline:
      fullnameOverride: vmalert
      server:
        env:
          - name: GOGC
            value: "100"
        resources:
          requests:
            memory: 200Mi
          limits:
            memory: 200Mi
```

**Comment by Haleygo** [comment]
Hi @andersbulderbank , thanks for all the info!
I checked the profiles, and the issue appears to be a performance degradation rather than a memory leak since [v1.104.0](https://github.com/VictoriaMetrics/VictoriaMetrics/releases/tag/v1.104.0). We increased the default `-remoteWrite.maxQueueSize` tenfold, which now requires `~50Mi(before is ~5Mi)` of memory for the remoteWriteClient. This usage doesn't reflect itself on RSS due to how the operating system calculates it, but gradually catching up.
>FEATURE: [vmalert](https://docs.victoriametrics.com/vmalert): bump default values for sending data to remoteWrite.url: remoteWrite.maxQueueSize from 100_000 to 1_000_000, remoteWrite.maxBatchSize from 1_000 to 10_000, remoteWrite.concurrency from 1 to 4. The new settings should improve remote write performance of vmalert with default settings.

OFC, the memory increase is not indefinite, it should stabilize after some time(assuming vmalert does not get OOM during this period). This issue significantly impacts small vmalert setups, as it previously required only a few dozen MiB of memory, and this change nearly doubles that requirement.
The current workaround is to reduce `-remoteWrite.maxQueueSize` to smaller value, the helm values would be:
```
    releaseName: victoria-metrics-alert
    version: 0.12.3
    repo: https://victoriametrics.github.io/helm-charts/
    valuesInline:
      fullnameOverride: vmalert
      server:
        extraArgs:
          rule: "/config/alerts-*.yaml"
          remoteWrite.maxQueueSize: "100000"
```

**Comment by andersbulderbank** [comment]
@Haleygo Thanks for the update, feel free to close this issue when it suits you.

**Comment by f41gh7** [comment]
The bugfix was included into [v1.107.0](https://docs.victoriametrics.com/changelog/#v11070)
